azkaban-common/src/main/java/azkaban/container/models/PodTemplateMergeUtils.java

```diff
@@ -42,95 +45,180 @@ private PodTemplateMergeUtils() {
+      templateContainer.setImagePullPolicy(podSpecContainer.getImagePullPolicy());
+    }
+
+    // Override Image from the corresponding podSpecContainer.
+    if (null != podSpecContainer.getImage()) {
+      templateContainer.setImage(podSpecContainer.getImage());
+    }
+
+    // Override Resources from the corresponding podSpecContainer.
+    if (null != podSpecContainer.getResources()) {
+      templateContainer.setResources(podSpecContainer.getResources());
     }
+
+    // Merge Volume Mounts with podSpecContainerVolumeMounts overriding
+    // templateContainerVolumeMounts.
+    final List<V1VolumeMount> templateContainerVolumeMounts =
+        templateContainer.getVolumeMounts();
+    final List<V1VolumeMount> podSpecContainerVolumeMounts =
+        podSpecContainer.getVolumeMounts();
+    final List<V1VolumeMount> mergedContainerVolumeMounts =
+        getMergedContainerVolumeMounts(
+            templateContainerVolumeMounts, podSpecContainerVolumeMounts);
+    templateContainer.setVolumeMounts(mergedContainerVolumeMounts);
   }
 
   /**
-   * Add those initContainers which are not already available in the podSpec
+   * This method merges VolumeMounts such that: 1) All VolumeMounts from the
+   * podSpecContainerVolumeMounts are added. 2) Only those VolumeMounts of
+   * templateContainerVolumeMounts are added which are not already part of
+   * podSpecContainerVolumeMounts.
    *
-   * @param podSpec     Already created podSpec using the {@link AzKubernetesV1SpecBuilder}
-   * @param podTemplate Instance of the class {@link AzKubernetesV1PodTemplate} to extract items
+   * @param templateContainerVolumeMounts List of VolumeMounts from templateContainer.
+   * @param podSpecContainerVolumeMounts  List of VolumeMounts from podSpecContainer.
+   * @return List of VolumeMounts after merging templateInitContainerVolumeMounts and
+   * podSpecInitContainerVolumeMounts.
    */
-  private static void mergeInitContainers(V1PodSpec podSpec,
-      AzKubernetesV1PodTemplate podTemplate) {
-    List<V1Container> podSpecInitContainers = podSpec.getInitContainers();
-    if (null != podSpecInitContainers) {
-      List<V1Container> templateInitContainers = podTemplate.getInitContainers(
-          tempInitContainer -> podSpecInitContainers.stream().map(V1Container::getName)
-              .noneMatch(name -> name.equals(tempInitContainer.getName())));
-      for (V1Container containerItem : templateInitContainers) {
-        podSpec.addInitContainersItem(containerItem);
-      }
+  private static List<V1VolumeMount> getMergedContainerVolumeMounts(
+      List<V1VolumeMount> templateContainerVolumeMounts,
+      List<V1VolumeMount> podSpecContainerVolumeMounts) {
+    final List<V1VolumeMount> allContainerVolumeMounts = new ArrayList<>();
+    if (null != templateContainerVolumeMounts && null != podSpecContainerVolumeMounts) {
+      List<V1VolumeMount> templateOnlyInitContainerVolumeMounts = templateContainerVolumeMounts
```

Comment:
This is very inefficient way to merge. Can you create a set from templateContainerVolumeMounts and iterate over podSpecContainerVolumentMounts to create union which would have anything common overridden from podSpec.
Convert the set into a list. Unless ofc, I am not able to understand this logic.

Reply:
This method is doing exactly the same. It is finding all the volume mounts in templateContainerVolumeMounts that are not part of podSpecContainerVolumentMounts. And then adding them all. 
It is however not necessary to create a list, hence updated it to utilize foreach loop.